src/state/apis/swapper/getUsdRatesApi.ts

```diff
@@ -0,0 +1,84 @@
+import type { AssetId } from '@shapeshiftoss/caip'
+import type { GetTradeQuoteInput } from '@shapeshiftoss/swapper'
+import type { GetTradeQuoteInputArgs } from 'components/Trade/hooks/useSwapper/getTradeQuoteArgs'
+import { getTradeQuoteArgs } from 'components/Trade/hooks/useSwapper/getTradeQuoteArgs'
+import { getBestSwapperApi } from 'state/apis/swapper/getBestSwapperApi'
+import { getUsdRateApi } from 'state/apis/swapper/getUsdRateApi'
+import { swapperApi } from 'state/apis/swapper/swapperApi'
+import type { State } from 'state/apis/types'
+
+export type GetUsdRatesArgs = {
+  feeAssetId: AssetId
+} & (
+  | { tradeQuoteInputArgs?: never; tradeQuoteArgs: GetTradeQuoteInput }
+  | { tradeQuoteInputArgs: GetTradeQuoteInputArgs; tradeQuoteArgs?: never }
+)
+
+type GetUsdRatesReturn = {
+  buyAssetUsdRate: string
+  sellAssetUsdRate: string
+  feeAssetUsdRate: string
+}
+
+const getBestSwapperType = getBestSwapperApi.endpoints.getBestSwapperType
+const getUsdRate = getUsdRateApi.endpoints.getUsdRate
+
+export const getUsdRatesApi = swapperApi.injectEndpoints({
+  endpoints: build => ({
+    getUsdRates: build.query<GetUsdRatesReturn, GetUsdRatesArgs>({
+      queryFn: async (args, { getState, dispatch }) => {
+        const { feeAssetId } = args
+        const buyAssetId = args.tradeQuoteInputArgs
+          ? args.tradeQuoteInputArgs.buyAsset.assetId
+          : args.tradeQuoteArgs.buyAsset.assetId
+        const sellAssetId = args.tradeQuoteInputArgs
+          ? args.tradeQuoteInputArgs.sellAsset.assetId
+          : args.tradeQuoteArgs.sellAsset.assetId
+        const state: State = getState() as unknown as State // ReduxState causes circular dependency
+        const { assets } = state
+        try {
+          const tradeQuoteArgs = args.tradeQuoteInputArgs
+            ? await getTradeQuoteArgs(args.tradeQuoteInputArgs)
+            : args.tradeQuoteArgs
+
+          if (!tradeQuoteArgs) throw new Error('tradeQuoteArgs is undefined')
+          const feeAsset = assets.byId[feeAssetId]
+          if (!feeAsset) throw new Error(`Asset not found for AssetId ${feeAssetId}`)
+
+          const swapperType = await dispatch(
+            getBestSwapperType.initiate({
+              ...tradeQuoteArgs,
+              feeAsset,
+            }),
+          ).then(r => r.data)
+
+          if (!swapperType) throw new Error(`Swapper type not found.`)
+
+          const feeAssetUsdRate = await dispatch(
+            getUsdRate.initiate({ assetId: feeAssetId, swapperType }),
+          ).then(r => r.data)
+          const buyAssetUsdRate = await dispatch(
+            getUsdRate.initiate({ assetId: buyAssetId, swapperType }),
+          ).then(r => r.data)
+          const sellAssetUsdRate = await dispatch(
+            getUsdRate.initiate({ assetId: sellAssetId, swapperType }),
+          ).then(r => r.data)
```

Comment:
parallelize?

Reply:
🧠

Reply:
Actioned in https://github.com/shapeshift/web/pull/3549/commits/b51512291d20d3c9a6895652cd305dd54fb9ece2.